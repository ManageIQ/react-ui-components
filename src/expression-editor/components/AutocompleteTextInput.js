import React from 'react';
import PropTypes from 'prop-types';
import { Button, Dropdown, DropdownToggle, DropdownItem, DropdownSeparator, TextInput } from '@patternfly/react-core';

import ExpressionEditorPropTypes from './ExpressionEditorPropTypes';
import { keyCodes } from "../constants";

export default class AutocompleteTextInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      toDelete: {},
      index: -1,
      isFocused: true
    };
  }

  onMenuClick = (index) => {
    const selected = this.props.options[index];
    this.setState({ index: -1 });
    this.props.onSubmit(selected);
  }

  onBlur = () => {
    this.setState({isFocused: false});
    this.handleSubmit(this.props.value);
  }

  onFocus = () => {
    this.setState({isFocused: true});
  }

  handleKeyDown = (e) => {
    const { target: { value } } = e;
    let index = 0;
    let dropdown;
    let dropdownItem;
    let itemPossition;
    switch (e.keyCode)  {
      case keyCodes.enter:
        this.handleSubmit(value);
        break;
      case keyCodes.upArrow:
        index = this.state.index <= -1 ? this.state.index : this.state.index - 1;
        dropdown = document.getElementsByClassName("pf-c-dropdown__menu")[0];
        dropdownItem = document.getElementsByClassName("pf-c-dropdown__menu-item")[0];
        itemPossition = dropdownItem.getBoundingClientRect().height * index;
        dropdown.scrollTo(0, itemPossition);
        this.setState(prevState => ({ index }));
        break;
      case keyCodes.downArrow:
        index = this.state.index >= this.props.options.length - 1 ? this.props.options.length - 1 : this.state.index + 1;
        dropdown = document.getElementsByClassName("pf-c-dropdown__menu")[0];
        dropdownItem = document.getElementsByClassName("pf-c-dropdown__menu-item")[0];
        itemPossition = dropdownItem.getBoundingClientRect().height * index;
        dropdown.scrollTo(0, itemPossition);

        this.setState(prevState => ({ index }));
        break;
      case keyCodes.leftArrow:
        if (value === '') {
          this.props.onKeyDown(e);
        }
        break;
      case keyCodes.end:
        this.props.onKeyDown(e);
        break;
      case keyCodes.home:
        this.props.onKeyDown(e);
        break;
      default:
        this.setState({ index: -1 });
    }
  }


  eligibleForSubmit = (value) => {
    if (value === "" && this.state.index === -1) {
      return false;
    }
    else if (this.props.aliasMode) {
      return value;
    }
    let selected = {label: ""};
    if (this.state.index === -1) {
      selected = this.props.options.find(this.props.matchingFunction(value)) || {
        id: value, label: value, type: 'userinput', next: this.props.next,
      };
      if (this.props.denyUserInput && selected.type === "userinput") {
        return false;
      }
    } else {
      selected = this.props.options[this.state.index];
    }
    if (selected.label === "") {
      return false;
    }
    return selected;

  }


  handleSubmit = (value) => {
    const selected = this.eligibleForSubmit(value);
    if (selected) {
      this.props.onSubmit(selected);
    }
    this.setState({ index: -1 });
  }

  addButtonClick = () => {
    this.handleSubmit(this.props.value)
  }

  handleChange = ({ target: { value } }) => {
    this.props.onChange(value);
  }

  generateMenuClick = (i) => () => {
    this.onMenuClick(i);
  }

  render() {
    return (
      <span className="maxDropdownHeght">
          <Dropdown
            toggle={
              <input
                autoFocus
                className="pf-c-form-control autocompleteTextInput"
                input="true"
                type="text"
                ref={this.props.inputRef}
                onKeyDown={this.handleKeyDown}
                value={this.props.value}
                onBlur={this.onBlur}
                onFocus={this.onFocus}
                fullWidth
                onChange={this.handleChange}
              />
            }
            isOpen={this.state.isFocused}
          >
            {this.props.options.map((o, i) => (
              <DropdownItem
                component="button"
                className="dropdownFontSize"
                isHovered={this.state.index === i}
                onMouseDown={this.generateMenuClick(i)}
              >
                {o.label}
              </DropdownItem>))}
          </Dropdown>

          <Button
            className="inputButton"
            onClick={this.addButtonClick}
            id="expressionEditorInputButton"
            isDisabled={!this.eligibleForSubmit(this.props.value)}
          >
           Add
          </Button>
      </span>

    );
  }
}

AutocompleteTextInput.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  submitKeyCode: PropTypes.number,
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onKeyDown: PropTypes.func.isRequired,
  deleteExpression: PropTypes.func,
  inputRef: PropTypes.object.isRequired,
  options: PropTypes.arrayOf(ExpressionEditorPropTypes.option).isRequired,
  next: PropTypes.arrayOf(ExpressionEditorPropTypes.term),
  matchingFunction: PropTypes.func,
  denyUserInput: PropTypes.bool,
};

AutocompleteTextInput.defaultProps = {
  value: '',
  matchingFunction: value => option => option.label.toLowerCase().includes(value.toLowerCase()),
};
