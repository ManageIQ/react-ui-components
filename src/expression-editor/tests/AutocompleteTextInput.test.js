import React from 'react';
import AutocompleteTextInput from '../components/AutocompleteTextInput';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Dropdown } from '@patternfly/react-core';

// TODO test internal method like eligibleForSubmit


describe('AutocompleteTextInput Component', () => {
  let dropdownItem;
  let dropdown;
  beforeAll(() => {
    dropdown = document.createElement('div');
    dropdown.className = "pf-c-dropdown__menu";
    dropdownItem = document.createElement('div');
    dropdownItem.className = "pf-c-dropdown__menu-item";
    dropdown.appendChild(dropdownItem);
    document.body.appendChild(dropdown);
  });

  afterAll(() => {
    document.body.removeChild(dropdown);
  });

  it('Calls onSubmit when custom user input is allowed', () => {
    let onSubmit = jest.fn();
    let onKeyDown = jest.fn();
    const wrapper = mount(
      <AutocompleteTextInput
        onSubmit={onSubmit}
        onChange={jest.fn()}
        onKeyDown={onKeyDown}
        options={[]}
        value={'ahoj'}
        denyUserInput={true}
        setAliasMode={jest.fn()}
        aliasMode={jest.fn()}
        isLastElement={false}
        deleteExpression={jest.fn()}
        blurAllChips={jest.fn()}
        inputRef={React.createRef()}
        index={0}
        next={[]}/>
      );

      // expect(component.eligibleForSubmit("")).toBe(true);
      wrapper.find('input').simulate('keydown', {key: 'Enter', keyCode: 13});
      expect(onKeyDown.mock.calls.length).toBe(0);
      expect(onKeyDown.mock.calls.length).toBe(0);
  });

  it('Does NOT call onSubmit when custom user input is allowed', () => {
    let onSubmit = jest.fn();
    let onKeyDown = jest.fn();
    const wrapper = mount(
      <AutocompleteTextInput
        onSubmit={onSubmit}
        onChange={jest.fn()}
        onKeyDown={onKeyDown}
        options={[]}
        value={'ahoj'}
        denyUserInput={false}
        setAliasMode={jest.fn()}
        aliasMode={jest.fn()}
        isLastElement={false}
        deleteExpression={jest.fn()}
        blurAllChips={jest.fn()}
        inputRef={React.createRef()}
        index={0}
        next={[]}/>
      );

      // expect(component.eligibleForSubmit("")).toBe(true);
      wrapper.find('input').simulate('keydown', {key: 'Enter', keyCode: 13});
      expect(onKeyDown.mock.calls.length).toBe(0);
      expect(onSubmit).toBeCalled();
  });

    it('Selects on submit first options by keyboard', () => {
    let options =  [{ id: 111, label: 'first', type: 'operator'}, { id: 112, label: 'second', type: 'operator'}];
    let onSubmit = jest.fn();
    let onKeyDown = jest.fn();
    console.log(JSON.stringify(document.getElementsByClassName("pf-c-dropdown__menu")));
    console.log(JSON.stringify(document.getElementsByClassName("pf-c-dropdown__menu-item")));
    const wrapper = mount(
      <AutocompleteTextInput
        onSubmit={onSubmit}
        onChange={jest.fn()}
        onKeyDown={onKeyDown}
        options={options}
        value={'ahoj'}
        denyUserInput={false}
        setAliasMode={jest.fn()}
        aliasMode={jest.fn()}
        isLastElement={false}
        deleteExpression={jest.fn()}
        blurAllChips={jest.fn()}
        inputRef={React.createRef()}
        index={0}
        next={[]}/>
      );

      // expect(component.eligibleForSubmit("")).toBe(true);
      // wrapper.find('button').at(1).simulate('click');
      wrapper.find('input').simulate('keydown', {keyCode: 40});
      wrapper.find('input').simulate('keydown', {key: 'Enter', keyCode: 13});
      expect(onKeyDown.mock.calls.length).toBe(0);
      expect(onSubmit).toBeCalled();
  });


  it('Selects on submit first options by pressing down -> up -> down key', () => {
  let options =  [{ id: 111, label: 'first', type: 'operator'}, { id: 112, label: 'second', type: 'operator'}];
  let onSubmit = jest.fn();
  let onKeyDown = jest.fn();
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={onSubmit}
      onChange={jest.fn()}
      onKeyDown={onKeyDown}
      options={options}
      value={'ahoj'}
      denyUserInput={false}
      setAliasMode={jest.fn()}
      aliasMode={jest.fn()}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={0}
      next={[]}/>
    );

    // expect(component.eligibleForSubmit("")).toBe(true);
    // wrapper.find('button').at(1).simulate('click');
    wrapper.find('input').simulate('keydown', {keyCode: 40});
    wrapper.find('input').simulate('keydown', {keyCode: 38});
    wrapper.find('input').simulate('keydown', {keyCode: 40});
    wrapper.find('input').simulate('keydown', {key: 'Enter', keyCode: 13});
    expect(onKeyDown.mock.calls.length).toBe(0);
    expect(onSubmit).toBeCalled();
  });

  it('Call onKeyDown for Home, End, leftArrow with empty value', () => {
  let onKeyDown = jest.fn();
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={jest.fn()}
      onChange={jest.fn()}
      onKeyDown={onKeyDown}
      options={[]}
      value={''}
      denyUserInput={false}
      setAliasMode={jest.fn()}
      aliasMode={jest.fn()}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={0}
      next={[]}/>
    );

    // expect(component.eligibleForSubmit("")).toBe(true);
    // wrapper.find('button').at(1).simulate('click');
    wrapper.find('input').simulate('keydown', {keyCode: 35});
    wrapper.find('input').simulate('keydown', {keyCode: 36});
    wrapper.find('input').simulate('keydown', {keyCode: 37});
    wrapper.find('input').simulate('keydown', {keyCode: 1});
    expect(wrapper.state().index).toBe(-1);
    expect(onKeyDown.mock.calls.length).toBe(3);
  });

  it('Selects on submit first options by pressing down -> up -> down key', () => {
  let onKeyDown = jest.fn();
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={jest.fn()}
      onChange={jest.fn()}
      onKeyDown={onKeyDown}
      options={[]}
      value={'ahoj'}
      denyUserInput={false}
      setAliasMode={jest.fn()}
      aliasMode={jest.fn()}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={0}
      next={[]}/>
    );

    // expect(component.eligibleForSubmit("")).toBe(true);
    // wrapper.find('button').at(1).simulate('click');

    wrapper.find('input').simulate('keydown', {keyCode: 37});
    expect(onKeyDown.mock.calls.length).toBe(0);
  });

  it('Selects on submit first options by mouse click', () => {
  let options =  [{ id: 111, label: 'first', type: 'operator'}, { id: 112, label: 'second', type: 'operator'}];
  let onSubmit = jest.fn();
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={onSubmit}
      onChange={jest.fn()}
      onKeyDown={jest.fn()}
      options={options}
      value={''}
      denyUserInput={true}
      setAliasMode={jest.fn()}
      aliasMode={jest.fn()}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={0}
      next={[]}/>
    );

    // expect(component.eligibleForSubmit("")).toBe(true);
    wrapper.find('button').at(1).simulate('mouseDown');
    // expect(onSubmit.mock.calls[0][0]).toBe({ id: 111, label: 'first', type: 'operator'});
    expect(onSubmit).toHaveBeenCalledWith({ id: 112, label: 'second', type: 'operator'});
});

  it('OnBlur calls submit and sets state', () => {
  let onSubmit = jest.fn();
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={onSubmit}
      onChange={jest.fn()}
      onKeyDown={jest.fn()}
      options={[]}
      value={'ahoj'}
      denyUserInput={false}
      setAliasMode={jest.fn()}
      aliasMode={jest.fn()}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={0}
      next={[]}/>
    );

    // expect(component.eligibleForSubmit("")).toBe(true);
    // wrapper.find('button').at(1).simulate('click');
    wrapper.find('input').simulate('focus');
    expect(wrapper.state().isFocused).toBeTruthy();
    wrapper.find('input').simulate('blur');
    expect(wrapper.state().isFocused).toBeFalsy();
    expect(onSubmit).toBeCalled();
  });

  it('onChange is called', () => {
  let onChange = jest.fn();
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={jest.fn()}
      onChange={onChange}
      onKeyDown={jest.fn()}
      options={[]}
      value={'ahoj'}
      denyUserInput={false}
      setAliasMode={jest.fn()}
      aliasMode={jest.fn()}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={0}
      next={[]}/>
    );

    // expect(component.eligibleForSubmit("")).toBe(true);
    // wrapper.find('button').at(1).simulate('click');
    wrapper.find('input').simulate('change', 'a');
    expect(wrapper.state().isFocused).toBeTruthy();
    expect(onChange).toBeCalled();
  });

  it('eligibleForSubmit return false with empty string argument', () => {
  let onChange = jest.fn();
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={jest.fn()}
      onChange={onChange}
      onKeyDown={jest.fn()}
      options={[]}
      value={'ahoj'}
      denyUserInput={false}
      setAliasMode={jest.fn()}
      aliasMode={jest.fn()}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={0}
      next={[]}/>
    );

    let spy = jest.spyOn(wrapper.instance(), 'eligibleForSubmit');
    let ret = wrapper.instance().eligibleForSubmit('');
    expect(spy).toHaveBeenCalled();
    expect(ret).toBe(false);
  });

  it('eligibleForSubmit return value with empty string argument and aliasMode=true', () => {
  let onChange = jest.fn();
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={jest.fn()}
      onChange={onChange}
      onKeyDown={jest.fn()}
      options={[]}
      value={'ahoj'}
      denyUserInput={false}
      setAliasMode={jest.fn()}
      aliasMode={true}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={1}
      next={[]}/>
    );

    let spy = jest.spyOn(wrapper.instance(), 'eligibleForSubmit');
    let ret = wrapper.instance().eligibleForSubmit('val');
    expect(spy).toHaveBeenCalled();
    expect(ret).toBe('val');
  });

  it('eligibleForSubmit returns matching options', () => {
  let options =  [{ id: 111, label: 'first', type: 'operator'}, { id: 112, label: 'second', type: 'operator'}];
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={jest.fn()}
      onChange={jest.fn()}
      onKeyDown={jest.fn()}
      options={options}
      value={'ahoj'}
      denyUserInput={false}
      setAliasMode={jest.fn()}
      aliasMode={false}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={-1}
      next={[]}/>
    );

    let spy = jest.spyOn(wrapper.instance(), 'eligibleForSubmit');
    let ret = wrapper.instance().eligibleForSubmit('fi');
    expect(spy).toHaveBeenCalled();
    expect(ret).toEqual({"id": 111, "label": "first", "type": "operator"});
  });

  it('eligibleForSubmit returns new userinput type object', () => {
  let options =  [{ id: 111, label: 'first', type: 'operator'}, { id: 112, label: 'second', type: 'operator'}];
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={jest.fn()}
      onChange={jest.fn()}
      onKeyDown={jest.fn()}
      options={options}
      value={'ahoj'}
      denyUserInput={false}
      setAliasMode={jest.fn()}
      aliasMode={false}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={-1}
      next={[]}/>
    );

    let spy = jest.spyOn(wrapper.instance(), 'eligibleForSubmit');
    let ret = wrapper.instance().eligibleForSubmit('bla');
    expect(spy).toHaveBeenCalled();
    expect(ret).toEqual({"id": "bla", "label": "bla", "next": [], "type": "userinput"});
  });

  it('eligibleForSubmit returns false if there are no matching options and user input is denied', () => {
  let options =  [{ id: 111, label: 'first', type: 'operator'}, { id: 112, label: 'second', type: 'operator'}];
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={jest.fn()}
      onChange={jest.fn()}
      onKeyDown={jest.fn()}
      options={options}
      value={'ahoj'}
      denyUserInput={true}
      setAliasMode={jest.fn()}
      aliasMode={false}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={-1}
      next={[]}/>
    );

    let spy = jest.spyOn(wrapper.instance(), 'eligibleForSubmit');
    let ret = wrapper.instance().eligibleForSubmit('bla');
    expect(spy).toHaveBeenCalled();
    expect(ret).toEqual(false);
  });

  it('eligibleForSubmit returns object on selected index', () => {
  let options =  [{ id: 111, label: 'first', type: 'operator'}, { id: 112, label: 'second', type: 'operator'}];
  const wrapper = mount(
    <AutocompleteTextInput
      onSubmit={jest.fn()}
      onChange={jest.fn()}
      onKeyDown={jest.fn()}
      options={options}
      value={'ahoj'}
      denyUserInput={true}
      setAliasMode={jest.fn()}
      aliasMode={false}
      isLastElement={false}
      deleteExpression={jest.fn()}
      blurAllChips={jest.fn()}
      inputRef={React.createRef()}
      index={1}
      next={[]}/>
    );

    wrapper.setState({index: 1});
    let spy = jest.spyOn(wrapper.instance(), 'eligibleForSubmit');
    let ret = wrapper.instance().eligibleForSubmit('');
    expect(spy).toHaveBeenCalled();
    expect(ret).toEqual({"id": 112, "label": "second", "type": "operator"});
    expect(wrapper.state().index).toBe(1);
  });

  it('Matches the snapshot', () => {
    const component = shallow(
      <AutocompleteTextInput
        onSubmit={jest.fn()}
        onChange={jest.fn()}
        onKeyDown={jest.fn()}
        options={[]}
        value={''}
        denyUserInput={true}
        setAliasMode={jest.fn()}
        aliasMode={jest.fn()}
        isLastElement={false}
        deleteExpression={jest.fn()}
        blurAllChips={jest.fn()}
        inputRef={React.createRef()}
        index={0}
        next={[]}/>
      );

      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
  });
})
