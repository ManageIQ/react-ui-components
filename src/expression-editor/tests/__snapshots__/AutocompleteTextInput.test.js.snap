// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteTextInput Component Matches the snapshot 1`] = `
<span
  className="maxDropdownHeght"
>
  <Dropdown
    className=""
    direction="down"
    dropdownItems={Array []}
    isOpen={true}
    isPlain={false}
    onSelect={[Function]}
    position="left"
    toggle={
      <input
        autoFocus={true}
        className="pf-c-form-control autocompleteTextInput"
        fullWidth={true}
        input="true"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    }
  />
  <Button
    aria-label={null}
    className="inputButton"
    component="button"
    id="expressionEditorInputButton"
    isActive={false}
    isBlock={false}
    isDisabled={true}
    isFocus={false}
    isHover={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Add
  </Button>
</span>
`;
